### PR TITLE
Add endpoints to export AuditEvents as unstructured data

### PR DESCRIPTION
**Pull request by tigrato**
This PR adds a new gRPC service `AuditLogService` that allows clients to request audit log events as unstructured data. Data is returned as `structpb.Struct` representation which is convertible to JSON.

Unstructured data allows clients to receive audit events without prior knowledge of their proto-wire representation if their end goal is to export the events as JSON to an upstream service. It allows clients whose version is older than Auth's version to receive events that their proto version doesn't support. Exporters can continue operating without losing events or having to upgrade them each time Auth is upgraded.

Part of https://github.com/gravitational/teleport-plugins/pull/821 
Part of https://github.com/gravitational/teleport/issues/23388

### CODE REVIEW

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+service AuditLogService {
+  // StreamUnstructuredSessionEvents streams audit events from a given session recording in an unstructured format.
+  // This endpoint is used by the event handler to retrieve the session events as JSON.
+  rpc StreamUnstructuredSessionEvents(StreamUnstructuredSessionEventsRequest) returns (stream EventUnstructured);
+  // GetUnstructuredEvents gets events from the audit log in an unstructured format.
+  // This endpoint is used by the event handler to retrieve the events as JSON.
+  rpc GetUnstructuredEvents(GetUnstructuredEventsRequest) returns (EventsUnstructured);
+}
+
+// StreamUnstructuredSessionEventsRequest is a request containing needed data to fetch a session recording.
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:34]:
```suggestion
// StreamUnstructuredSessionEventsRequest is a request containing data needed to fetch a session recording.
```

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+message StreamUnstructuredSessionEventsRequest {
+  // session_id is the ID for a given session in an UUIDv4 format.
+  string session_id = 1;
+  // StartIndex is the index of the event to resume the stream after.
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:38]:
```suggestion
  // start_index is the index of the event to resume the stream after.
```

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  // A start_index of 0 creates a new stream.
+  int32 start_index = 2;
+}
+
+// Order specifies any ordering of some objects as returned in regards to some aspect
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:43]: Not sure about this one but I think it should be "an" and not "any"?

```suggestion
// Order specifies an ordering of some objects as returned in regards to some aspect
```

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+// of said objects which may be trivially ordered such as a timestamp.
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:44]: This sentence is a bit hard to follow. Since `Order` seems to be specific to the order of events by the timestamp, could we make this comment refer to that specific use case? I think it'd make the comment easier to understand than the current abstract version.

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  // set to its value. Otherwise leave empty.
+  string start_key = 6;
+  // order specifies an ascending or descending order of events.
+  // A value of 0 means a descending order and a value of 1 means an ascending order.
+  Order order = 7;
+}
+
+// EventsUnstructured represents a list of events.AuditEvent in unstructured format.
+message EventsUnstructured {
+  // Items is a list of unstructured formatted audit events.
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:73]:
```suggestion
  // items is a list of unstructured formatted audit events.
```

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+// Order specifies any ordering of some objects as returned in regards to some aspect
+// of said objects which may be trivially ordered such as a timestamp.
+enum Order {
+  ORDER_DESCENDING_UNSPECIFIED = 0;
+  ORDER_ASCENDING = 1;
+}
+
+// GetUnstructuredEventsRequest is a request with the needed data to fetch events.
+message GetUnstructuredEventsRequest {
+  // namespace, if not set, defaults to 'default'
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:52]: Nit but the comments for fields of `GetUnstructuredEventsRequest` should probably end with a dot as well, just to keep it consistent across comments in this file. ;)

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  // order specifies an ascending or descending order of events.
+  // A value of 0 means a descending order and a value of 1 means an ascending order.
+  Order order = 7;
+}
+
+// EventsUnstructured represents a list of events.AuditEvent in unstructured format.
+message EventsUnstructured {
+  // Items is a list of unstructured formatted audit events.
+  repeated EventUnstructured items = 1;
+  // the key of the last event if the returned set did not contain all events found i.e limit <
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:75]:
```suggestion
  // last_key is the key of the last event if the returned set did not contain all events found i.e limit <
```

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  // actual amount. this is the key clients can supply in another API request to continue fetching
+  // events from the previous last position
+  string last_key = 2;
+}
+
+// EventUnstructured represents a siggle events.AuditEvent in an unstructured format.
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:81]:
```suggestion
// EventUnstructured represents a single events.AuditEvent in an unstructured format.
```

File: api/proto/teleport/auditlog/v1/auditlog.proto [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+  // start_key is used to resume a query in order to enable pagination.
+  // If the previous response had LastKey set then this should be
+  // set to its value. Otherwise leave empty.
+  string start_key = 6;
+  // order specifies an ascending or descending order of events.
+  // A value of 0 means a descending order and a value of 1 means an ascending order.
+  Order order = 7;
+}
+
+// EventsUnstructured represents a list of events.AuditEvent in unstructured format.
```
Comment (ravicious) [inline api/proto/teleport/auditlog/v1/auditlog.proto:71]:
```suggestion
// EventsUnstructured represents a list of events.AuditEvent in an unstructured format.
```

File: api/proto/buf.yaml [hunk 10]
```diff
@@ -27,6 +27,8 @@ lint:
     - teleport/legacy/types/types.proto
     - teleport/legacy/types/wrappers/wrappers.proto
   ignore_only:
+    RPC_NO_SERVER_STREAMING:
+      - teleport/auditlog/v1/auditlog.proto
```
Comment (ravicious) [inline api/proto/buf.yaml:31]: Did you consider adding that to the ignore item for `UNARY_RPC` below?

OTOH, `RPC_NO_SERVER_STREAMING` is more specific than `UNARY_RPC`. But `UNARY_RPC` already has a comment as to why `UNARY_RPC` is used in the first place. 🤔

**Review by ravicious — APPROVED** [review]
Submitted some suggestions for grammar/structure changes of the proto comments.

File: api/proto/buf.yaml (hunk 10 above)
Comment (tigrato) [reply api/proto/buf.yaml]: I missed that. Nice catch.

File: lib/auth/grpcserver.go [hunk 11]
```diff
@@ -5144,3 +5147,68 @@ func (g *GRPCServer) authenticate(ctx context.Context) (*grpcContext, error) {
...
+		unstructuredEvents = append(unstructuredEvents, unstructuredEvent)
+	}
+
+	return &auditlogpb.EventsUnstructured{
+		Items:   unstructuredEvents,
+		LastKey: lastkey,
+	}, nil
+}
+
+// StreamUnstructuredSessionEventsServer streams all events from a given session recording as an unstructured format. An error is returned on the first
```
Comment (AntonAM) [inline lib/auth/grpcserver.go:5185]: I think "error is returned on the first channel... " part of the comment from the original streaming function is not suited for this one, we have different signature.

**Review by AntonAM — COMMENTED** [review]
A few questions.

**Review by AntonAM — APPROVED** [review]
